pkg/storage/storage.go

```diff
@@ -80,69 +84,121 @@ func (p *storage) GetContainerMetrics(pods ...apitypes.NamespacedName) ([]api.Ti
 
 	timestamps := make([]api.TimeInfo, len(pods))
 	resMetrics := make([][]metrics.ContainerMetrics, len(pods))
-
-	for i, pod := range pods {
-		metricPoint, present := p.pods[pod]
+	for podIdx, pod := range pods {
+		contPoints, present := p.pods[pod]
 		if !present {
 			continue
 		}
 
-		contMetrics := make([]metrics.ContainerMetrics, len(metricPoint.Containers))
-		var earliestTS *time.Time
-		for i, contPoint := range metricPoint.Containers {
-			contMetrics[i] = metrics.ContainerMetrics{
+		contMetrics := make([]metrics.ContainerMetrics, len(contPoints))
+		var earliestTS time.Time
+		var contIdx int
+		for contName, contPoint := range contPoints {
+			prevContPoint := p.prevPods[pod][contName]
+			cpuUsage := cpuUsageOverTime(contPoint.MetricsPoint, prevContPoint.MetricsPoint)
+			contMetrics[contIdx] = metrics.ContainerMetrics{
 				Name: contPoint.Name,
 				Usage: corev1.ResourceList{
-					corev1.ResourceName(corev1.ResourceCPU):    contPoint.CpuUsage,
+					corev1.ResourceName(corev1.ResourceCPU):    cpuUsage,
 					corev1.ResourceName(corev1.ResourceMemory): contPoint.MemoryUsage,
 				},
 			}
-			if earliestTS == nil || earliestTS.After(contPoint.Timestamp) {
-				ts := contPoint.Timestamp // copy to avoid loop iteration variable issues
-				earliestTS = &ts
+			if earliestTS.IsZero() || earliestTS.After(contPoint.Timestamp) {
+				earliestTS = contPoint.Timestamp
 			}
+			contIdx++
 		}
-		if earliestTS == nil {
-			// we had no containers
-			earliestTS = &time.Time{}
-		}
-		timestamps[i] = api.TimeInfo{
-			Timestamp: *earliestTS,
+		resMetrics[podIdx] = contMetrics
+		timestamps[podIdx] = api.TimeInfo{
+			Timestamp: earliestTS,
 			Window:    kubernetesCadvisorWindow,
 		}
-		resMetrics[i] = contMetrics
 	}
 	return timestamps, resMetrics, nil
 }
 
 func (p *storage) Store(batch *MetricsBatch) {
-	newNodes := make(map[string]NodeMetricsPoint, len(batch.Nodes))
-	var nodeCount, containerCount int
-	for _, nodePoint := range batch.Nodes {
+	p.storeNodeMetrics(batch.Nodes)
+	p.storePodMetrics(batch.Pods)
+}
+
+func (p *storage) storeNodeMetrics(nodes []NodeMetricsPoint) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	prevNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	var nodeCount int
+	for _, nodePoint := range nodes {
 		if _, exists := newNodes[nodePoint.Name]; exists {
 			klog.Errorf("duplicate node %s received", nodePoint.Name)
 			continue
 		}
-		nodeCount += 1
 		newNodes[nodePoint.Name] = nodePoint
+		nodeCount++
+
+		if nodePoint.Timestamp.After(p.nodes[nodePoint.Name].Timestamp) {
```

Comment:
Can this really happen in normal situation? I think this can only help if:
1. cadvisor reported time is broken (clock stopped, or totally broke)
2. scrapes were stored in reversed order

For first case I think it's reasonable to assume that we don't need to handle it and just reject such data and log error.
For second case I don't think we should ensure that scrapes are stored in correct order instead making storage order aware.

What do you think?

Reply:
This check is intended to guard from scraping the same metric point between two scrapes. This happens often and we don't want to update the previous metric point if that's the case.

> For second case I don't think we should ensure that scrapes are stored in correct order instead making storage order aware.

I am not too sure what you mean by that. Would you have any suggestions regarding how the metric points are stored?

Reply:
@serathius could you please update me regarding the comment above?

Reply:
ok, make sense. Good catch.
Could you leave a comment explaining this.

Reply:
done in the latest changes